src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs

```diff
@@ -177,41 +193,51 @@ private unsafe int EncodeIntoBuffer(char* buffer, int bufferLength, char* value,
                     *buffer = firstChar;
                     buffer++;
                     bufferLength--;
-                    totalWritten++;
+                    charsWritten++;
                 }
                 else
                 {
-                    int nextScalar = UnicodeHelpers.GetScalarValueFromUtf16(firstChar, secondChar, out wasSurrogatePair);
-                    if (!TryEncodeUnicodeScalar(nextScalar, buffer, bufferLength, out charsWritten))
+                    int nextScalar = UnicodeHelpers.GetScalarValueFromUtf16(firstChar, secondChar, out wasSurrogatePair, out bool _);
+                    if (!TryEncodeUnicodeScalar(nextScalar, buffer, bufferLength, out int charsWrittenThisTime))
                     {
-                        throw new ArgumentException("Argument encoder does not implement MaxOutputCharsPerInputChar correctly.");
+                        charsConsumed = (int)(originalBuffer - buffer);
+                        return OperationStatus.DestinationTooSmall;
                     }
 
-                    buffer += charsWritten;
-                    bufferLength -= charsWritten;
-                    totalWritten += charsWritten;
                     if (wasSurrogatePair)
                     {
                         secondCharIndex++;
                     }
+
+                    buffer += charsWrittenThisTime;
+                    bufferLength -= charsWrittenThisTime;
+                    charsWritten += charsWrittenThisTime;
                 }
             }
 
             if (secondCharIndex == valueLength)
             {
                 firstChar = value[valueLength - 1];
-                int nextScalar = UnicodeHelpers.GetScalarValueFromUtf16(firstChar, null, out wasSurrogatePair);
-                if (!TryEncodeUnicodeScalar(nextScalar, buffer, bufferLength, out charsWritten))
+                int nextScalar = UnicodeHelpers.GetScalarValueFromUtf16(firstChar, null, out wasSurrogatePair, out bool wasUnmatchedSurrogate);
+                if (!isFinalBlock && wasUnmatchedSurrogate)
```

Comment:
I believe this check is incorrect. `NeedMoreData` should only be returned if `isFinalBlock` is _false_ and the very last character of the buffer is a UTF-16 high surrogate character.

Reply:
I believe the code is doing what you state. `!isFinalBlock` + `wasUnmatchedSurrogate`.

Reply:
Won't _wasUnmatchedSurrogate_ also be true if the very last character of the buffer is a UTF-16 low surrogate character?

Reply:
Yes thanks. Changed the variable name to make it more clear.

Reply:
Just to clarify - I don't have a strong preference on what the variable name is called. :)

All I wanted to check was that if _isFinalBlock_ is _false_ and the very last char is a high surrogate character, the API returns `NeedsMoreData`. Any other combination (_isFinalBlock_ is _true, or the very last char is a low surrogate character) should not return `NeedsMoreData`. I didn't see a unit test for this condition so I wanted to call it out as something to double-check.

Reply:
I renamed `wasUnmatchedSurrogate` to `needMoreData` and changed the logic so `needMoreData` returns true\false as you state.